### PR TITLE
Block export on IRM

### PR DESCRIPTION
**Pull request by jbloom22**

### CODE REVIEW

File: src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala [hunk 1]
```diff
@@ -687,3 +688,134 @@ private class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix)
...
+      val nColsInBlock = if (blockColIndex == truncatedBlockCol) excessCols else blockSize
+      var offset = 0
+      blockRowDependencies(blockRowIndex).foreach { parent =>
+        val firstRowInParent =  parentPartitionBoundaries(parent)
+        val lastRowInParent = parentPartitionBoundaries(parent + 1) // technically, first row in next IRM partition
+
+        val firstIndexToCopy = math.max(0, firstRowInBlock - firstRowInParent).toInt
+        val lastIndexToCopy = (math.min(lastRowInBlock, lastRowInParent) - firstRowInParent).toInt
+        
+        val indexedRowsInParent = irm.rows.iterator(parentPartitions(parent), context)
```
Comment (cseed) [inline src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala]: You're computing the parent once for each block.  You should write out all the blocks at once.

File: src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala [hunk 2]
```diff
@@ -687,3 +688,134 @@ private class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix)
...
+  override def getDependencies: Seq[Dependency[_]] = {  
+    Array[Dependency[_]](
+      new NarrowDependency(irm.rows) {
+        def getParents(partitionId: Int): Seq[Int] = blockRowDependencies(partitionId)
+      }
+    )
+  }
+  
+  protected def getPartitions: Array[Partition] =
+    (0 until gp.rowPartitions).map(IntPartition).toArray[Partition]
```
Comment (cseed) [inline src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala]: The "traditional" approach is to have a `WriteBlocksRDDPartition` that has the relevant information necessary to compute the partition, and then use that in `compute` and `getDependencies`.  See `BlockedRDD`, for example.  That avoids having to call `computeBlockRowDependencies` multiple times.

File: src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala [hunk 3]
```diff
@@ -687,3 +688,134 @@ private class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix)
...
+        var i = 0
+        while (i < firstIndexToCopy) {
+          indexedRowsInParent.next()
+          i += 1
+        }
+        while (i < lastIndexToCopy) {
+          val indexedRow = indexedRowsInParent.next()
+          assert(indexedRow.index == firstRowInParent + i)
+                    
+          System.arraycopy(indexedRow.vector.toArray, firstColInBlock, data0, offset, nColsInBlock)
```
Comment (cseed) [inline src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala]: This should be free of copies.

File: src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala [hunk 4]
```diff
@@ -687,3 +688,134 @@ private class BlockMatrixMultiplyRDD(l: BlockMatrix, r: BlockMatrix)
...
+          
+          offset += nColsInBlock
+          i += 1
+        }
+
+        val data =
+          if (blockColIndex == truncatedBlockCol)
+            data0.take(nColsInBlock * nRowsInBlock)
```
Comment (cseed) [inline src/main/scala/is/hail/distributedmatrix/BlockMatrix.scala]: This is also a copy.

**Comment by jbloom22** [comment]
Will re-open in separate PR.

**Pull request closed without merging**